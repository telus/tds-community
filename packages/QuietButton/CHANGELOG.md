# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [0.2.0](https://github.com/telus/tds-community/compare/@tds/community-quiet-button@0.1.3...@tds/community-quiet-button@0.2.0) (2021-09-21)


### Features

* **community-quiet-button:** react 17 and styled components 5 ([12229ce](https://github.com/telus/tds-community/commit/12229ce5ded1a423f765237b7ac618910752e86e))





## [0.1.3](https://github.com/telus/tds-community/compare/@tds/community-quiet-button@0.1.2...@tds/community-quiet-button@0.1.3) (2020-11-02)


### Bug Fixes

* **community-quiet-button:** address issue [#362](https://github.com/telus/tds-community/issues/362), fix bug ([d09fb28](https://github.com/telus/tds-community/commit/d09fb282a6ecf70a0c05a18fcfccd5efce3395b4))





## [0.1.2](https://github.com/telus/tds-community/compare/@tds/community-quiet-button@0.1.1...@tds/community-quiet-button@0.1.2) (2020-09-09)


### Bug Fixes

* **community-quiet-button:** update util-helpers version ([4789296](https://github.com/telus/tds-community/commit/4789296c77ce434c129c41b5f227bf3410b14f4c))





## [0.1.1](https://github.com/telus/tds-community/compare/@tds/community-quiet-button@0.1.0...@tds/community-quiet-button@0.1.1) (2020-06-19)

**Note:** Version bump only for package @tds/community-quiet-button





# 0.1.0 (2020-06-12)


### Features

* create QuietButton component ([e8ee35e](https://github.com/telus/tds-community/commit/e8ee35e73bd23978f2c8aebfe7ed62176ca9f4f2))
