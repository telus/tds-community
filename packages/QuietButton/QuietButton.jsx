import React from 'react'
import styled from 'styled-components'
import Box from '@tds/core-box'
import {
  colorGreyRaven,
  colorWhite,
  colorGreyShuttle,
  colorGreyGainsboro,
  colorGreyShark,
} from '@tds/core-colours'
import { componentWithName } from '@tds/util-prop-types'
import { safeRest } from '@tds/util-helpers'
import { small } from '@tds/shared-typography'
import { borders } from '@tds/shared-styles'
import PropTypes from 'prop-types'

const baseButton = {
  boxSizing: 'border-box',
  margin: '0.125rem',
  cursor: 'pointer',
  background: colorWhite,
  transition: 'all 0.2s ease-in-out',
  minWidth: '2.75rem',
  minHeight: '1.75rem',
  border: `0.0625rem solid ${colorGreyRaven}`,
  position: 'relative',
  borderRadius: '0.1875rem',
  color: colorGreyShark,
  '&:hover': {
    boxShadow: `0 0 0 0.125rem ${colorGreyRaven}`,
    midWidth: '4.5rem',
  },
  '&:active': {
    border: `0.0625rem solid ${colorGreyShuttle}`,
    boxShadow: `0 0 0 0.125rem ${colorGreyShuttle}`,
    background: colorGreyGainsboro,
  },
  '&:focus': {
    background: colorWhite,
    boxShadow: `0 0 0 0.125rem ${colorGreyRaven}, 0 0 0 0.125rem ${colorWhite} inset, 0 0 0 0.1875rem ${colorGreyShuttle} inset`,
    outline: 'none !important',
  },
  '@media (prefers-reduced-motion: reduce)': {
    transition: 'none !important',
  },
}
const btnWrapper = {
  display: 'flex',
  flexDirection: 'row',
  alignItems: 'flex-start',
  '& svg': {
    margin: '-0.125rem 0rem -0.3125rem 0rem',
  },
}
const spaceWrapper = {
  paddingRight: '0.25rem',
}

const StyledBox = styled(Box)`
  padding-top: 0.125rem;
  padding-bottom: 0.3125rem;
`

const StyledQuietButton = styled.button(baseButton, small, borders.rounded)
const ButtonWrapper = styled.div(btnWrapper)
const SpaceWrapper = styled.div(spaceWrapper)

const spaceFunction = childrenArray => {
  return childrenArray.map((child, index) => {
    if (child.type && child.type.name !== 'A11yContent' && index === childrenArray.length - 1) {
      return child
    }
    return <SpaceWrapper key={child || child.type.name}>{child}</SpaceWrapper>
  })
}
/**
 * The quiet button is used for optional actions, and only comes in one variant and size
 * @version ./package.json
 * @visibleName QuietButton (beta)
 */

const QuietButton = ({ children, ...rest }) => {
  const childrenArray = Array.isArray(children) ? children : [children]

  return (
    <StyledQuietButton type="button" {...safeRest(rest)}>
      <StyledBox horizontal={2}>
        <ButtonWrapper>{spaceFunction(childrenArray)}</ButtonWrapper>
      </StyledBox>
    </StyledQuietButton>
  )
}

QuietButton.propTypes = {
  /**
   * Button children.
   */

  children: PropTypes.oneOfType([
    PropTypes.arrayOf(
      PropTypes.oneOfType([
        PropTypes.string,
        componentWithName('A11yContent'),
        componentWithName('Dependent', true),
      ])
    ),
    PropTypes.string,
  ]).isRequired,
}

export default QuietButton
