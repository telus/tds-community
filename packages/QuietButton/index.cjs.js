const QuietButton = require('./dist/index.cjs')

module.exports = QuietButton
