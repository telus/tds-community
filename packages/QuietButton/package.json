{
  "name": "@tds/community-quiet-button",
  "version": "0.1.1",
  "description": "",
  "main": "index.cjs.js",
  "module": "index.es.js",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/telus/tds-community.git"
  },
  "publishConfig": {
    "access": "public"
  },
  "author": "TELUS digital",
  "engines": {
    "node": ">=8"
  },
  "bugs": {
    "url": "https://github.com/telus/tds-community/issues"
  },
  "homepage": "http://tds.telus.com",
  "peerDependencies": {
    "react": "^16.8.2",
    "react-dom": "^16.8.2",
    "styled-components": "^4.1.3"
  },
  "dependencies": {
    "@tds/core-box": "^2.2.0",
    "@tds/core-colours": "^2.2.1",
    "@tds/shared-styles": "^1.5.2",
    "@tds/shared-typography": "^1.3.5",
    "@tds/util-helpers": "^1.2.0",
    "@tds/util-prop-types": "^1.3.2",
    "prop-types": "^15.6.2"
  }
}
