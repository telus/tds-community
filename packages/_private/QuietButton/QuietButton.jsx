import React from 'react'
import styled from 'styled-components'
import { safeRest } from '@tds/util-helpers'
import { small } from '@tds/shared-typography'
import { borders } from '@tds/shared-styles'
import PropTypes from 'prop-types'

const baseButton = {
  boxSizing: 'border-box',
  margin: '2px',
  padding: '0px 16px 0px 16px',
  cursor: 'pointer',
  background: '#FFFFFF',
  transition: 'all 0.2s ease-in-out',
  minWidth: '44px',
  height: '28px',
  border: '1px solid #71757B',
  position: 'relative',
  borderRadius: '3px',
  color: '#2A2C2E',
  '&:hover': {
    boxShadow: '0 0 0 2px #71757B',
    midWidth: '72px',
  },
  '&:active': {
    border: '1px solid #54595F',
    boxShadow: '0 0 0 2px #54595F',
    background: '#D8D8D8',
  },
  '&:focus': {
    background: '#FFFFFF',
    boxShadow: '0 0 0 2px #71757B,0 0 0 2px #FFFFFF inset, 0 0 0 3px #54595F inset',
    outline: 'none !important',
  },
  '@media (prefers-reduced-motion: reduce)': {
    transition: 'none !important',
  },
}
const btnWrapper = {
  display: 'flex',
  flexDirection: 'row',
  alignItems: 'center',
  padding: '3px 0px 5px 0px',
  '& svg': {
    margin: '-3px 0 -5px 0',
  },
}
const spaceWrapper = {
  paddingRight: '4px',
}

const StyledQuietButton = styled.button(baseButton, small, borders.rounded)
const ButtonWrapper = styled.div(btnWrapper)
const SpaceWrapper = styled.div(spaceWrapper)

const spaceFunction = childrenArray => {
  return childrenArray.map((child, index) => {
    if (child.type && child.type.name !== 'A11yContent' && index === childrenArray.length - 1) {
      return child
    }
    return <SpaceWrapper key={child || child.type.name}>{child}</SpaceWrapper>
  })
}
/**
 * The quiet button is used for optional actions, and only comes in one variant and size
 * @version ./package.json
 */

const QuietButton = ({ children, ...rest }) => {
  const childrenArray = Array.isArray(children) ? children : [children]

  return (
    <StyledQuietButton type="button" {...safeRest(rest)}>
      <ButtonWrapper>{spaceFunction(childrenArray)}</ButtonWrapper>
    </StyledQuietButton>
  )
}

QuietButton.propTypes = {
  /**
   * Button children.
   */

  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]).isRequired,
}

export default QuietButton
