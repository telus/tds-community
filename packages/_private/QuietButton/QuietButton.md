Quiet buttons are used sparingly and for an optional action. Mainly used in internal applications like My TELUS (web and app) or Casa; it only comes in one variant and size.

### Usage criteria

- Use for interactions that are not necessarily mandatory
- Use for enhanced experience where if the button does not exist, the customer can still do what they need to
- Keep the text short and precise, recommendation is text that describes an action (Ie. Copy, Edit, Print)
- Make use of the A11yContent core component to provide more written context for assistive technology users for when they navigate a page using only button landmarks
- Buttons should not be disabled

```jsx
<QuietButton>Text</QuietButton>
```

#### QuietButton with Icons

- Only dependent icons can be added into buttons, they can be on the left or right of the label
- The colour of the text and icon have to be the same

```jsx
<QuietButton>
  <Edit />
  Left Icon
</QuietButton>
```

```jsx
<QuietButton>
  Right Icon
  <Print />
</QuietButton>
```

#### Using A11yContent

Use the `A11yContent` component to create invisible text that is read out loud by screen readers.

```jsx
<QuietButton>
  <A11yContent>testing</A11yContent>
  With A11y Content
</QuietButton>
```

#### Adding Functionality

Provide a function as the onClick prop to perform an action when clicked.

```jsx
<QuietButton onClick={() => alert('You clicked the button!')}>Click me!</QuietButton>
```
