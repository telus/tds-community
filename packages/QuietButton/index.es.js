import QuietButton from './dist/index.es'

export default QuietButton
