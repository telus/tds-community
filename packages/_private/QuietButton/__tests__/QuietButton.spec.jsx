import React from 'react'
import { shallow, mount } from 'enzyme'
import 'jest-styled-components'

import A11yContent from '@tds/core-a11y-content'
import { Edit } from '@tds/core-interactive-icon'
import QuietButton from '../QuietButton'

describe('QuietButton', () => {
  const doShallow = (props = {}) => shallow(<QuietButton {...props}>Words</QuietButton>)
  const doMount = (props = {}) => mount(<QuietButton {...props}>Words</QuietButton>)

  it('renders', () => {
    const quietButton = doMount()

    expect(quietButton).toMatchSnapshot()
    quietButton.unmount()
  })

  it('does not allow custom CSS', () => {
    const quietButton = doShallow({
      className: 'my-custom-class',
      style: { color: 'hotpink' },
    })

    expect(quietButton).not.toHaveProp('className', 'my-custom-class')
    expect(quietButton).not.toHaveProp('style')
  })

  it('should contain A11yContent', () => {
    const button = shallow(
      <QuietButton>
        Go home
        <A11yContent>testing</A11yContent>
      </QuietButton>
    )

    expect(button).toContainReact(<A11yContent>testing</A11yContent>)
  })

  it('should contain an interactiveIcon', () => {
    const button = shallow(
      <QuietButton>
        Go home
        <Edit />
      </QuietButton>
    )

    expect(button).toContainReact(<Edit />)
  })

  it('should contain text', () => {
    const button = shallow(<QuietButton>Text</QuietButton>)

    expect(button.text()).toEqual('Text')
  })

  it('should do something when clicked', () => {
    const mockCallBack = jest.fn()
    const button = shallow(<QuietButton onClick={mockCallBack}>Text</QuietButton>)
    button.simulate('click')
    expect(mockCallBack.mock.calls.length).toEqual(1)
  })
})
