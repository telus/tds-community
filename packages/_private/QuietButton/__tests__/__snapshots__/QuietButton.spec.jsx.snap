// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuietButton renders 1`] = `
.c0 {
  box-sizing: border-box;
  margin: 2px;
  padding: 0px 16px 0px 16px;
  cursor: pointer;
  background: #FFFFFF;
  -webkit-transition: all 0.2s ease-in-out;
  transition: all 0.2s ease-in-out;
  min-width: 44px;
  height: 28px;
  border: 1px solid #71757B;
  position: relative;
  border-radius: 3px;
  color: #2A2C2E;
  word-wrap: break-word;
  font-size: 0.875rem;
  -webkit-letter-spacing: -0.6px;
  -moz-letter-spacing: -0.6px;
  -ms-letter-spacing: -0.6px;
  letter-spacing: -0.6px;
  line-height: 1.42857;
  border-radius: 4px;
}

.c0:hover {
  box-shadow: 0 0 0 2px #71757B;
  mid-width: 72px;
}

.c0:active {
  border: 1px solid #54595F;
  box-shadow: 0 0 0 2px #54595F;
  background: #D8D8D8;
}

.c0:focus {
  background: #FFFFFF;
  box-shadow: 0 0 0 2px #71757B,0 0 0 2px #FFFFFF inset,0 0 0 3px #54595F inset;
  outline: none !important;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 3px 0px 5px 0px;
}

.c1 svg {
  margin: -3px 0 -5px 0;
}

.c2 {
  padding-right: 4px;
}

@media (prefers-reduced-motion:reduce) {
  .c0 {
    -webkit-transition: none !important;
    transition: none !important;
  }
}

<QuietButton>
  <QuietButton__StyledQuietButton
    type="button"
  >
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "QuietButton__StyledQuietButton-sc-1anctzy-0",
            "isStatic": false,
            "lastClassName": "c0",
            "rules": Array [
              "box-sizing: border-box;",
              "margin: 2px;",
              "padding: 0px 16px 0px 16px;",
              "cursor: pointer;",
              "background: #FFFFFF;",
              "transition: all 0.2s ease-in-out;",
              "min-width: 44px;",
              "height: 28px;",
              "border: 1px solid #71757B;",
              "position: relative;",
              "border-radius: 3px;",
              "color: #2A2C2E;",
              "&:hover {",
              "box-shadow: 0 0 0 2px #71757B;",
              "mid-width: 72px;",
              "}",
              "&:active {",
              "border: 1px solid #54595F;",
              "box-shadow: 0 0 0 2px #54595F;",
              "background: #D8D8D8;",
              "}",
              "&:focus {",
              "background: #FFFFFF;",
              "box-shadow: 0 0 0 2px #71757B,0 0 0 2px #FFFFFF inset, 0 0 0 3px #54595F inset;",
              "outline: none !important;",
              "}",
              "@media (prefers-reduced-motion: reduce) {",
              "transition: none !important;",
              "}",
              "word-wrap: break-word;",
              "font-size: 0.875rem;",
              "letter-spacing: -0.6px;",
              "line-height: 1.42857;",
              "border-radius: 4px;",
            ],
          },
          "displayName": "QuietButton__StyledQuietButton",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "QuietButton__StyledQuietButton-sc-1anctzy-0",
          "target": "button",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
      type="button"
    >
      <button
        className="c0"
        type="button"
      >
        <QuietButton__ButtonWrapper>
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "QuietButton__ButtonWrapper-sc-1anctzy-1",
                  "isStatic": false,
                  "lastClassName": "c1",
                  "rules": Array [
                    "display: flex;",
                    "flex-direction: row;",
                    "align-items: center;",
                    "padding: 3px 0px 5px 0px;",
                    "& svg {",
                    "margin: -3px 0 -5px 0;",
                    "}",
                  ],
                },
                "displayName": "QuietButton__ButtonWrapper",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "QuietButton__ButtonWrapper-sc-1anctzy-1",
                "target": "div",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
          >
            <div
              className="c1"
            >
              <QuietButton__SpaceWrapper
                key="Words"
              >
                <StyledComponent
                  forwardedComponent={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "attrs": Array [],
                      "componentStyle": ComponentStyle {
                        "componentId": "QuietButton__SpaceWrapper-sc-1anctzy-2",
                        "isStatic": false,
                        "lastClassName": "c2",
                        "rules": Array [
                          "padding-right: 4px;",
                        ],
                      },
                      "displayName": "QuietButton__SpaceWrapper",
                      "foldedComponentIds": Array [],
                      "render": [Function],
                      "styledComponentId": "QuietButton__SpaceWrapper-sc-1anctzy-2",
                      "target": "div",
                      "toString": [Function],
                      "warnTooManyClasses": [Function],
                      "withComponent": [Function],
                    }
                  }
                  forwardedRef={null}
                >
                  <div
                    className="c2"
                  >
                    Words
                  </div>
                </StyledComponent>
              </QuietButton__SpaceWrapper>
            </div>
          </StyledComponent>
        </QuietButton__ButtonWrapper>
      </button>
    </StyledComponent>
  </QuietButton__StyledQuietButton>
</QuietButton>
`;
